mod_papers/helper.php

```diff
@@ -13,150 +14,257 @@
-            }
-            //Print results
-            if ($work['parse'] === 1) {
-                $output .= '<b>' . $work['work-title']['title']['value'] . '</b><br>';
-                // $doi = $ids['work-external-identifier-id']['value'];
-
-                if (strcmp($work['work-citation']['work-citation-type'], 'BIBTEX') == 0) {
-                    $bibtex = $work['work-citation']['citation'];
-                    $volume = '';
-                    $pages  = '';
-                    if (preg_match('/volume\\s?=\\s?{(\\d+)}/', $bibtex, $match)) {
-                        $volume = $match[1];
-                    }
-                    if (preg_match('/pages\\s?=\\s?{([0-9-]+)}/', $bibtex, $match)) {
-                        $pages = $match[1];
-                    }
-                }
-
-                if (!is_null($work['work-contributors'])) {
-                    foreach ($work['work-contributors']['contributor'] as $authors) {
-                        if (($authors === reset($work['work-contributors']['contributor'])) && ($authors === end($work['work-contributors']['contributor']))) {
-                            $output .= $authors['credit-name']['value'] . '<br>';
-                        } elseif ($authors === reset($work['work-contributors']['contributor'])) {
-                            $output .= $authors['credit-name']['value'];
-                        } elseif ($authors === end($work['work-contributors']['contributor'])) {
-                            $output .= ' and ' . $authors['credit-name']['value'] . '<br>';
-                        } else {
-                            $output .= ', ' . $authors['credit-name']['value'];
-                        }
-                    }
-                } else {
-                    //Get authorlist from bibtex
-                    if (preg_match('/author\\s?=\\s?{(.+)}/', $bibtex, $match)) {
-                        $authorstr = $match[1];
-                        $authors   = explode(" and ", $authorstr);
-                        foreach ($authors as $author) {
-                            if (($author === reset($authors)) && ($author === end($authors))) {
-                                $output .= $author . '<br>';
-                            } elseif ($author === reset($authors)) {
-                                $output .= $author;
-                            } elseif ($author === end($authors)) {
-                                $output .= ' and ' . $author . '<br>';
-                            } else {
-                                $output .= ', ' . $author;
-                            }
-                        }
-                    }
-                }
-
-                $output .= '<a href="http://dx.doi.org/' . $work['udoi'] . '">' . $work['journal-title']['value'] . ' <b>' . $volume . '</b> ' . $pages . ' (' . $work['publication-date']['year']['value'] . ')</a><br style="line-height:2.5em;">';
-            }
-        }
-
-        return $output;
-    }
+	/**
```

Comment:
Thanks for looking into this! Your diff has basically written me a new file here though, so it'll take a bit to digest properly.

Reply:
I apologize for such massive refactoring. I should have made several pull requests while the changes are progressing. My first attempt was to change as little as possible since I am not very familiar with php. But eventually it turned out that implementing two-step sequential downloads is very slow, almost a minute (it might be that i just did it wrong), so I started finding solutions how to speed the code up.

Reply:
As an option we can make pull request in an another branch, not master.